src/core/MIGRATION.md

```diff
@@ -933,6 +934,66 @@ For a few plugins, some of these steps (such as angular removal) could be a mont
 
 One convention that is useful for this is creating a dedicated `public/np_ready` directory to house the code that is ready to migrate, and gradually move more and more code into it until the rest of your plugin is essentially empty. At that point, you'll be able to copy your `index.ts`, `plugin.ts`, and the contents of `./np_ready` over into your plugin in the new platform, leaving your legacy shim behind. This carries the added benefit of providing a way for us to introduce helpful tooling in the future, such as [custom eslint rules](https://github.com/elastic/kibana/pull/40537), which could be run against that specific directory to ensure your code is ready to migrate.
 
+## Keep Kibana fast
+**tl;dr**: Load as much code lazily as possible.
+Everyone loves snappy applications with responsive UI and hates spinners. Users deserve the best user experiences regardless of whether they run Kibana locally or in the cloud, regardless of their hardware & environment.
```

Comment:
It's outside of code splitting, but another thing that is currently increasing our bundles size drastically are static imports that can be done between plugins (at least until https://github.com/elastic/kibana/pull/62390 is merged)

See https://github.com/elastic/kibana/pull/61700/commits/43a9d456eac807b405c2e4e7ea47e1f6048b5999 for example, where even with async mount method for the route mount function, all dependencies were kept because of a static import from another plugin that was using them (in addition to shipping the whole dependency plugin).

Not sure if it belongs there or more in our `CONVENTION`, but we should probably also declare that static exports from plugins should be avoided as much as possible and exposed from the plugins contracts instead. This ensure both that plugins will not import more that they should from other plugins and also reduce the overall initial load, as the 'static' code will only be shipped in the providing plugin code.

Reply:
> but we should probably also declare that static exports from plugins should be avoided as much as possible and exposed from the plugins contracts instead

I do agree it's a problem: a plugin importing anything from the `data` plugin will get additional (mostly unnecessary) 1.4Mb of code.
But I'd argue we shouldn't ban of static imports usage. Using static imports is a valid use-case for sharing stateless code. If we forced plugins to use plugins contracts only, it would worsen DX because it increases the number of dependencies passed down the code.  I'd rather say the current situation is a **technical** limitation, that shouldn't dictate plugin authors how to write their code and platform & operation teams  **have to** fix it eventually.

Reply:
After thinking a little about it you are right imho, this is only driven by a technical limitation that should be addressed, and is not a bad practice on itself.